pkg/ingester/tailer.go

```diff
@@ -0,0 +1,231 @@
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+// This is to limit size of dropped stream to make it easier for querier to query.
+// While dropping stream we divide it into batches and use start and end time of each batch to build dropped stream metadata
+const maxDroppedStreamSize = 10000
+
+type tailer struct {
+	id       uint32
+	orgID    string
+	matchers []*labels.Matcher
+	regexp   *regexp.Regexp
+
+	sendChan chan *logproto.Stream
+	closed   bool
+
+	blockedAt      *time.Time
+	blockedMtx     sync.RWMutex
+	droppedStreams []*logproto.DroppedStream
+
+	conn logproto.Querier_TailServer
+}
+
+func newTailer(orgID, query, regex string, conn logproto.Querier_TailServer) (*tailer, error) {
+	matchers, err := parser.Matchers(query)
+	if err != nil {
+		return nil, err
+	}
+
+	var re *regexp.Regexp
+	if regex != "" {
+		re, err = regexp.Compile(regex)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &tailer{
+		orgID:          orgID,
+		matchers:       matchers,
+		regexp:         re,
+		sendChan:       make(chan *logproto.Stream, 2),
+		conn:           conn,
+		droppedStreams: []*logproto.DroppedStream{},
+		id:             generateUniqueID(orgID, query, regex),
+	}, nil
+}
+
+func (t *tailer) loop() {
+	var stream *logproto.Stream
+	var err error
+
+	for {
+		if t.closed {
+			return
+		}
+
+		stream = <-t.sendChan
+		if stream == nil {
+			t.close()
+			return
+		}
+
+		// while sending new stream pop lined up dropped streams metadata for sending to querier
+		tailResponse := logproto.TailResponse{Stream: stream, DroppedStreams: t.popDroppedStreams()}
+		err = t.conn.Send(&tailResponse)
+		if err != nil {
+			level.Error(util.Logger).Log("Error writing to tail client", fmt.Sprintf("%v", err))
+			t.close()
+			return
+		}
+	}
+}
+
+func (t *tailer) send(stream logproto.Stream) {
+	if t.closed {
+		return
+	}
+
+	// if we are already dropping streams due to blocked connection, drop new streams directly to save some effort
+	if blockedSince := t.blockedSince(); blockedSince != nil {
+		if blockedSince.Before(time.Now().Add(-time.Second * 15)) {
+			t.close()
+			close(t.sendChan)
+			return
+		}
+		t.dropStream(stream)
+		return
+	}
+
+	if t.regexp != nil {
+		// filter stream by regex from query
+		t.filterEntriesInStream(&stream)
+		if len(stream.Entries) == 0 {
+			return
+		}
+	}
+
+	select {
+	case t.sendChan <- &stream:
+	default:
+		t.dropStream(stream)
+	}
+}
+
+func (t *tailer) filterEntriesInStream(stream *logproto.Stream) {
+	filteredEntries := new([]logproto.Entry)
+	for _, entry := range stream.Entries {
+		if t.regexp.MatchString(entry.Line) {
+			*filteredEntries = append(*filteredEntries, entry)
+		}
+	}
+
+	stream.Entries = *filteredEntries
+}
+
+func (t *tailer) isWatchingLabels(metric model.Metric) bool {
+	var labelValue model.LabelValue
+	var ok bool
+
+	for _, matcher := range t.matchers {
+		labelValue, ok = metric[model.LabelName(matcher.Name)]
```

Comment:
Not necessarily true. We can have a matcher match the "absence" of a label. This is a better example: https://github.com/cortexproject/cortex/blob/9c1a267d3e3fb392518f867b07a35cdc6c7de7df/pkg/chunk/chunk_store_utils.go#L32-L44